Mixin forwarder signature points to a nonexistent nested class owned by the subclass

## Compiler version

- 3.9.0-RC1
- 3.9.0-RC2
- 3.9.0-RC3
- 3.10.0-RC1-bin-20260712-f51c618-NIGHTLY

The same code works with 3.8.3, so the regression starts in 3.9.0-RC1.

## Minimized code

`Repro.scala`:

```scala
package repro

trait Base[S, D] {
  final class Helper
  final def helper(): Helper = new Helper
}

abstract class Impl[S, D] extends Base[S, D]
```

`Use.java`:

```java
package repro;

final class Use {
  void test(Impl<String, Integer> impl) {
    impl.helper();
  }
}
```

Compile with:

```shell
scala-cli compile Repro.scala -S 3.9.0-RC3 --server=false -d out
javac -cp out -d out Use.java
```

## Output

```text
Use.java:5: error: cannot access Impl$Helper
    impl.helper();
        ^
  class file for repro.Impl$Helper not found
```

`javap -p -s repro.Impl` shows:

```java
public final repro.Impl<S, D>.Helper helper();
  descriptor: ()Lrepro/Base$Helper;
```

The descriptor correctly refers to the class that exists, `Base$Helper`, while the generic `Signature` attribute refers to nonexistent `Impl$Helper`. With 3.8.3, the method is exposed as:

```java
public final repro.Base$Helper helper();
  descriptor: ()Lrepro/Base$Helper;
```

## Expectation

`Helper` is declared in `Base`, and the only generated nested class is `Base$Helper`. A mixin forwarder's generic signature must not relocate that class to `Impl`. It should either refer to `Base[S, D]#Helper` / `Base$Helper` consistently with the descriptor, or omit a redundant signature.

This breaks Java subclasses in [Apache Pekko PR #3303](https://github.com/apache/pekko/pull/3303): the compiler emits an `AbstractFSM[S,D].TransformHelper` return signature although the actual class is `FSM$TransformHelper`, and javac fails before the Java test can run. This appears related to generic-signature work around #24619 and #25880, but those issues do not cover this nonexistent-owner case.

--- a/compiler/src/dotty/tools/backend/jvm/GenericSignatures.scala
+++ b/compiler/src/dotty/tools/backend/jvm/GenericSignatures.scala
@@ -235,7 +235,16 @@ object GenericSignatures {
               boxedSig(tp.widenDealias.widenNullaryMethod)
         }
 
-      pre.widenDealias match {
+      // when generating a java generic signature that includes
+      // a selection of an inner class p.I, (p = `pre`, I = `cls`) must
+      // rewrite to p'.I, where p' refers to the class that directly defines
+      // the nested class I. see:
+      // https://github.com/scala/scala3/issues/26532
+      // https://github.com/scala/bug/issues/2585
+      val reboundPre =
+        if pre.exists then pre.baseType(sym.owner)
+        else pre
+      reboundPre.widenDealias match {
         // If the class is an inner class of a generic class, we must emit the outer generic class with its parameters
         // (see test `inner-of-generic` for an example of Java compatibility)
         case RefOrAppliedType(preSym: ClassSymbol, prePre, preArgs) if preArgs.nonEmpty =>

--- a/tests/generic-java-signatures/i26532.check
+++ b/tests/generic-java-signatures/i26532.check
@@ -1,0 +1,1 @@
+public final i26532.Base<S, D>$Helper i26532.Impl.helper()

--- a/tests/generic-java-signatures/i26532/Definitions_1.scala
+++ b/tests/generic-java-signatures/i26532/Definitions_1.scala
@@ -1,0 +1,9 @@
+package i26532
+
+trait Base[S, D]:
+  final class Helper:
+    def identity(value: S): S = value
+
+  final def helper(): Helper = new Helper
+
+abstract class Impl[S, D] extends Base[S, D]

--- a/tests/generic-java-signatures/i26532/Test_3.scala
+++ b/tests/generic-java-signatures/i26532/Test_3.scala
@@ -1,0 +1,6 @@
+import i26532.Impl
+
+object Test:
+  def main(args: Array[String]): Unit =
+    val helper = classOf[Impl[?, ?]].getDeclaredMethod("helper")
+    println(helper.toGenericString)

--- a/tests/generic-java-signatures/i26532/Use_2.java
+++ b/tests/generic-java-signatures/i26532/Use_2.java
@@ -1,0 +1,7 @@
+package i26532;
+
+final class Use_2 {
+    void test(Impl<String, Integer> impl) {
+        String value = impl.helper().identity("ok");
+    }
+}

